docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,302 @@
+can change over time, it is important to keep the checks in sync with the `default/main.yml`.
+The following checks can be generated:
+
+```yaml
+- name: "Fail if r_openshift_node_deployment_type is not defined"
+  fail:
+    msg: "r_openshift_node_deployment_type must be specified for this role"
+  when:
+  - r_openshift_node_deployment_type is not defined
+
+- name: Fail if invalid r_openshift_node_deployment_type provided
+  fail:
+    msg: "r_openshift_node_deployment_type can only be set to a single value from ['origin', 'openshift-enterprise']"
+  when:
+  - r_openshift_node_deployment_type is defined
+  - r_openshift_node_deployment_type not in ['origin', 'openshift-enterprise']
+```
+
+All the checks can be stored into `tasks/pre_checks.yml` file and included in the main task file.
+
+### Local variables
+
+All variables set inside a role has to be prefixed with `__` string.
+The only exception are role variables that are read outside of a role after
+the role finishes.
+
+### Role invocation
+
+Based on a role caller some role variables are always defined,
+some are defined only when corresponding inventory (or other global) variables are set.
+E.g. although `r_openshift_node_is_containerized` is set to `false` be default,
+the `openshift.common.is_containerized` is always set so it is safe to set the value unconditionally:
+```yaml
+- role: openshift_node
+  r_openshift_node_is_containerized: "{{ openshift.common.is_containerized }}"
+```
+
+Some role variables like `r_openshift_node_kubelet_args` are optionally defined
+and can not be set unless their corresponding inventory (or other playbook) variables
+are set. So the proposed way to set them is:
+```yaml
+- role: openshift_node
+  r_openshift_node_kubelet_args: "{{ openshift_node_kubelet_args | default(omit) }}"
+```
+
+Thus, in general the proposed way to invoke a role corresponds to the following pattern:
+```yaml
+- include_role:
+    name: <role>
+  vars:
+    r_<role>_<variable>_1: "{{ <global_variable>_1 | default(omit) }}"
+    ...
+    r_<role>_<variable>_j: "{{ <global_variable>_j | default(omit) }}"
+    r_<role>_<variable>_k: "{{ <global_variable>_k }}"
+    ...
+    r_<role>_<variable>_n: "{{ <global_variable>_n }}"
+```
+
+The `j` denotes a number of role variables that are set to non-role variables
+that does not have to be defined.
+The `n` denotes the total number of all role variables that are set.
+
+### Example
+
+**Before applying the proposal**:
+
+`openshift_node/tasks/main.yml`:
+```yaml
+...
+- name: pull node image
+  command: >
+    docker pull {{ openshift_node_image_name }}:{{ openshift_node_image_tag }}
+  register: docker_pull_result
+...
+```
+
+`playbooks/common/openshift-node/config.yml`
+```yaml
+...
+- role: openshift_node
+...
+```
+
+**After:**
+
+`openshift_node/defaults/main.yml`:
+```yaml
+__omit_str: "__{{ omit.split('__')[1] }}__"
```

Comment:
Absolutely. None of the `default` and `vars` is a good place for it as it is a hack. Would be great to make it part of the Ansible itself. Even better, make it a filter like `is_omitted` and used it as:
```yaml
- set_fact:
    r_openshift_node_kubelet_args: "{{ __openshift_node_kubelet_args }}"
  when: r_openshift_node_kube_args | is_omitted
```